net/freeradius/src/opnsense/mvc/app/controllers/OPNsense/Freeradius/common/OpenVpn.php

```diff
@@ -0,0 +1,213 @@
+                // thats a openvpn legacy tool to create CCDs for a specific server
+                // lets use this to ensure compatibility
+                $ccdConfigAsString = openvpn_csc_conf(self::ccdToLegacyStructure($ccd), $server);
+
+                self::writeCCDforServer($ccd->common_name, $ccdConfigAsString, $server['vpnid']);
+            }
+        }
+    }
+
+    /**
+     * Writes the ccd configuration we created using the legacy method to the disk at the correct location for a specific server
+     * @param string $common_name
+     * @param string $ccdConfigAsString
+     * @param string $openvpn_id
+     */
+    static function writeCCDforServer($common_name, $ccdConfigAsString, $openvpn_id)
+    {
+        openvpn_create_dirs();
+        // 'stolen' from openvpn_configure_csc - we cannot reuse this function since its not designed to
+        $target_filename = "/var/etc/openvpn-csc/{$openvpn_id}/{$common_name}";
+        file_put_contents($target_filename, $ccdConfigAsString);
+        chown($target_filename, 'nobody');
+        chgrp($target_filename, 'nobody');
+    }
+
+    /**
+     * This method is missing in the legacy API completely
+     * @param string $common_name
+     * @param string $openvpn_id
+     */
+    static function deleteCCDforServer($common_name, $openvpn_id)
+    {
+        $target_filename = "/var/etc/openvpn-csc/{$openvpn_id}/{$common_name}";
+        @unlink($target_filename);
+    }
+
+    static function ccdToLegacyStructure(CCD $ccd)
+    {
+        return (array)$ccd;
+    }
+
+    /**
+     * returns a CCD, which is based on your static(default) ccd and has been overlayed by the dynamic CCD
+     * @param CCD $static
+     * @param CCD $dynamic
+     * @return CCD
+     */
+    static function overlayCcd(CCD $static, CCD $dynamic)
+    {
+        // lets keep the reference intact
+        $ccd = clone($static);
+
+        $overridable_attribs = ['tunnel_network', 'tunnel_network6'];
+        foreach ($overridable_attribs as $attrib) {
+            if (isset($dynamic->{$attrib})) {
+                $ccd->{$attrib} = $dynamic->{$attrib};
+            }
+        }
+        return $ccd;
+    }
+
+    /**
+     * @return array an array of VPN-Servers ( stdClass ) which have the feature dynamic-ccd-lookup enabled
+     */
+    static function getDynamicCCDopenVPNServers()
+    {
+        $configObj = Config::getInstance()->object();
+        $servers = array();
+
+        if (isset($configObj->openvpn)) {
+            /** @var \SimpleXMLElement $root */
+            $root = $configObj->openvpn;
+            foreach ($root->children() as $name => $vpnServer) {
+                // if that VPN server has dynamic ccd enabled
+                if ($vpnServer->{'dynamic-ccd-lookup'} == '1') {
+                    // convert that one to an assoc array for easier usage in the toolchain
+                    $servers[$name] = json_decode(json_encode($vpnServer), true);
```

Comment:
really needed that way?

Reply:
json_decode(json_encode ?

Yes, if you hand over nested XML objects, thats by far the best generic method to convert into hashmaps or bjectmaps